logictest: delayed server shutdown

pkg/sql/logictest/tests/local/local_test.TestLogic_tenant [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14393028?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/14393028?buildTab=artifacts#/) on release-23.2 @ [47c0b1bb4862fa095a7ddb6ae0f8424b0276eaa8](https://github.com/cockroachdb/cockroach/commits/47c0b1bb4862fa095a7ddb6ae0f8424b0276eaa8):

Fatal error:

```
panic: test timed out after 14m55s
running tests:
	TestLogic_tenant (14m26s)

```
Stack: 

```
goroutine 38874 [running]:
testing.(*M).startAlarm.func1()
	GOROOT/src/testing/testing.go:2259 +0x320
created by time.goFunc
	GOROOT/src/time/sleep.go:176 +0x38
```
<details><summary>Log preceding fatal error</summary>
<p>

```
* github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start.func1()
* 	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:229 +0x34
* github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x4011195340?)
* 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x58
* github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start(0x4011195340, {0x6446130?, 0x4010ce17d0?})
* 	github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:228 +0x78
* github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start.func1()
* 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:120 +0x34
* github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError(0x4010480200?)
* 	github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:92 +0x58
* github.com/cockroachdb/cockroach/pkg/sql/colflow.(*FlowCoordinator).Start(0x4010480200, {0x64460f8?, 0x40111f85f0?})
* 	github.com/cockroachdb/cockroach/pkg/sql/colflow/flow_coordinator.go:119 +0x70
* github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBaseNoHelper).Run(0x4010480200, {0x64460f8?, 0x40111f85f0?}, {0x6416c78?, 0x4010476380})
* 	github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:725 +0x4c
* github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run(0x4010ec8500, {0x64460f8?, 0x40111f85f0}, 0x4?)
* 	github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:579 +0x180
* github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlow).Run(0x64af300?, {0x64460f8?, 0x40111f85f0?}, 0x40?)
* 	github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:305 +0x1d0
* github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run(0x40017ffa40, {0x6446130, 0x4011012810}, 0x400397e000, 0x4010510b40, 0x400ffeae00, 0x4010476380, 0x40086e6b40, 0x4010c7dae0)
* 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:905 +0x8d0
* github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun(0x40017ffa40, {0x6446130, 0x4011012810}, 0x40086e6b40, 0x400397e000, 0x420318?, {{0x6448dd0, 0x4003f3e138}, 0x0}, 0x4010476380, ...)
* 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1945 +0x170
* github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll.func3(0x983a440?, 0xffff81b364a0?, {0x6446130?, 0x4011012810?}, 0x4010a8dd60?, 0x10?, 0x10?)
* 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1660 +0xa0
* github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunAll(0x400397e000?, {0x6446130?, 0x4011012810}, 0x40086e6b40, 0x400397e000, 0x40086e6678, 0x4010476380, 0x0)
* 	github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1663 +0x180
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine(0x40086e6000, {0x6446130?, 0x4011012810}, 0x40086e6678, 0x0?, {0x64b0930?, 0x40111f8460}, 0x4010c7e090?, 0x40085cf718)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:2405 +0x320
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0x40086e6000, {0x6446130, 0x4011012030}, 0x40086e6678, {0x64b0930?, 0x40111f8460})
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1944 +0xe40
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0x40086e6000, {0x6446130, 0x4011012030}, {{0x6467a60, 0x400253e0e0}, {0x0, 0x0, 0x0}, {0x400efbe540, 0x28}, ...}, ...)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1125 +0x3714
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt.func1({0x6446130?, 0x401096fb30?})
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:145 +0x8c
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithProfiling(0x6446130?, {0x6446130?, 0x401096fb30?}, {0x6467a60?, 0x400253e0e0?}, 0x4f6559e?, 0x1a?)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:3403 +0x27c
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0x40086e6000, {0x6446130, 0x401096fb30}, {{0x6467a60, 0x400253e0e0}, {0x0, 0x0, 0x0}, {0x400efbe540, 0x28}, ...}, ...)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:144 +0x32c
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd.func1(0x4006b57c38, 0x4006b57850, 0x40086e6000, 0x4006b57778, 0x4009bcb870, 0x4009bcb860, 0x4009bcb880, 0x4006b57890)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2264 +0x2c0
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execCmd(0x40086e6000)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2269 +0x2a8
* github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0x40086e6000, {0x6446130, 0x401096f590}, 0x40042d9f38?, 0x40042d9fa8?, 0x0?)
* 	github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:2186 +0x168
* github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx.func1()
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:232 +0x90
* created by github.com/cockroachdb/cockroach/pkg/sql.(*InternalExecutor).runWithEx in goroutine 37807
* 	github.com/cockroachdb/cockroach/pkg/sql/internal.go:231 +0x134
* 
*
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_tenant.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-36718

--- a/pkg/kv/db.go
+++ b/pkg/kv/db.go
@@ -1177,7 +1177,9 @@ func getOneRow(runErr error, b *Batch) (KeyValue, error) {
 func IncrementValRetryable(ctx context.Context, db *DB, key roachpb.Key, inc int64) (int64, error) {
 	var err error
 	var res KeyValue
-	for r := retry.Start(base.DefaultRetryOptions()); r.Next(); {
+	retryOpts := base.DefaultRetryOptions()
+	retryOpts.Closer = db.Context().Stopper.ShouldQuiesce()
+	for r := retry.StartWithCtx(ctx, retryOpts); r.Next(); {
 		res, err = db.Inc(ctx, key, inc)
 		if errors.HasType(err, (*kvpb.UnhandledRetryableError)(nil)) ||
 			errors.HasType(err, (*kvpb.AmbiguousResultError)(nil)) {

